dask/dataframe/core.py

```diff
@@ -6061,9 +6063,16 @@ def map_partitions(
         graph = HighLevelGraph.from_collections(name, layer, dependencies=args)
         return Scalar(graph, name, meta)
     elif not (has_parallel_type(meta) or is_arraylike(meta) and meta.shape):
-        # If `meta` is not a pandas object, the concatenated results will be a
-        # different type
-        meta = make_meta(_concat([meta]), index=meta_index)
+        if meta_is_emulated:
+            # If `meta` is not a pandas object, the concatenated results will be a
+            # different type
+            meta = make_meta(_concat([meta]), index=meta_index)
+        else:
+            raise ValueError(
+                "Meta is not valid, `map_partitions` expects output to be a pandas object. "
+                "Try passing a pandas object as meta or a dict or tuple representing the "
+                "(name, dtype) of the columns."
+            )
```

Comment:
I have a couple questions to check my understanding!

1. if `meta_is_emulated is False & (has_parallel_type(meta) or is_arraylike(meta) and meta.shape)`, then we know `meta` is not valid? And that's why there's no need for an `elif meta_is_emulated is False` condition?
2. It seems there were already helpful errors raised for when inferring `meta` doesn't work if the user doesn't pass anything (with [`dataframe.utils.raise_on_meta_error`](https://github.com/dask/dask/blob/c1c88f066672c0b216fc24862a2b36a0a9fb4e22/dask/dataframe/utils.py#L165-L196)), but less so for when the user passes in something invalid?

Reply:
1. the `else` here is coming from `if meta_is_emulated` so it is equivalent to `elif meta_is_emulated is False`. Basically I am trying to say, if the meta is not a pandas object but dask tried to guess the meta (`meta_is_emulated is True`) then let's fix it. If on the other hand the meta is not a pandas object and the user provided it explicitly (`meta_is_emulated is False`) then raise an error.
2. Right. If you look at the original issue you can see the kind of confusing behavior that you get. Essentially the meta was reduced down to a single value (it was `1`) at this spot in the codebase, and then it was being coerced back into a series, so the index was totally lost.

Reply:
Ahh that makes sense, thank you! The other question I thought of is whether this is an issue elsewhere, but I think I've answered that one well enough for myself, though maybe not very efficiently. I went through the other methods with a `meta` argument and it seems like they either have a more strict enforcement of the meta type (e.g. in [`Series.map`](https://github.com/dask/dask/blob/c1c88f066672c0b216fc24862a2b36a0a9fb4e22/dask/dataframe/core.py#L3602) or rely on `map_partitions` (which will now be improved!)